consul/service.sls

```diff
@@ -4,41 +4,31 @@
 {#- Get the `tplroot` from `tpldir` #}
 {%- set tplroot = tpldir.split('/')[0] %}
 {%- from tplroot ~ "/map.jinja" import mapdata as consul with context %}
+{%- from tplroot ~ "/libtofs.jinja" import files_switch with context %}
 
 consul-init-env:
   file.managed:
-    {%- if grains['os_family'] == 'Debian' %}
-    - name: /etc/default/consul
-    {%- else %}
-    - name: /etc/sysconfig/consul
+    - name: {{ consul.service_env_path }}
     - makedirs: True
-    {%- endif %}
     - user: root
-    - group: root
+    - group: {{ consul.rootgroup }}
     - mode: '0644'
     - contents:
       - CONSUL_USER={{ consul.user }}
       - CONSUL_GROUP={{ consul.group }}
 
 consul-init-file:
   file.managed:
-    {%- if salt['test.provider']('service').startswith('systemd') %}
-    - source: salt://{{ tplroot }}/files/consul.service
-    - name: /etc/systemd/system/consul.service
+    - name: {{ consul.service_path }}
+    - source: {{ files_switch(['consul_service_unit.jinja'],
+                              'consul-init-file',
+                              ["id", "init", "os_family"]
+                 )
+              }}
```

Comment:
```suggestion
    - source: {{ files_switch(['consul_service_unit.jinja'],
                              lookup='consul-init-file',
                              default_files_switch=["id", "os_family", "init"]
                 )
              }}
```

* TOFS is difficult enough to understand as it is, so let's provide the argument names to keep things clearer.
* You'll see I've also changed the ordering to `["id", "os_family", "init"]` -- that's because it goes from most specific to most general (the opposite of the order in `map_jinja.yaml`).

Reply:
I have been thinking about this TOFS order:
What criteria do you use to conclude `init` is more general than `os_family` knowing we can use different `init` system within an `os_family` and same `init` system can be used on different `os_family`.
Instinctively, I think that `init` system is less general than `os_family` because it comes after the choice of `os_family` and `os` and if is effectively the case, we have to rethink about the order of `map_jinja.yaml`

Reply:
> What criteria do you use to conclude `init` is more general than `os_family` knowing we can use different `init` system within an `os_family` and same `init` system can be used on different `os_family`. Instinctively, I think that `init` system is less general than `os_family` because it comes after the choice of `os_family` and `os` and if is effectively the case, we have to rethink about the order of `map_jinja.yaml`

I'm not sure about this, in general:

1. software provides defaults in their code base
2. the OS could override these defaults at compile time

I think the same could be think of about `init` systems, in your PR, only `service_path` is configured depending of the `init` system, `parameter/init/` could disappear completely, see below.

> I've done this choice also because of the absence of `grains['init']` for `FreeBSD` and the layout of https://github.com/saltstack/salt/blob/master/salt/grains/core.py

You can symlink `consul/files/default/consul_service_unit.jinja` → `consul/files/sysvinit/consul_service_unit.jinja`, this way, when `grains[init]` does not exist we assume `sysvinit` as a default.

> What if, if we want specific files combining values of `init` grain and `os_family` grain. Let's say a different file for `sysVinit` on `Debian` and `sysVinit` on `gentoo`

The idea behind `TOFS` is to [limit the use of jinja](https://docs.saltproject.io/en/latest/topics/development/conventions/formulas.html#easy-on-the-jinja) by providing different files if we need to much logic between OS for the same file.

If there is a need, you can make `consul.service` include `consul.service.{{ grains.get("init", "sysvinit") }}` to dispatch the init system:
1. `consul.service.sysvinit` can
   1. use `files_switch(["consul_sysvinit.jinja"])`, then TOFS will select the file based on it's configuration (`["id", "os_family"]` by default)
   2. use a parameter like `sysv_init_file_path` which can be overwritten per `os_family` etc, the normal way
2. `consul.service.systemd` can
   1. use `files_switch(["consul_systemd.jinja"])`, then TOFS will select the file based on it's configuration (`["id", "os_family"]` by default)
   2. use a parameter like `systemd_init_file_path` which can be overwritten per `os_family` etc, the normal way
3. `consul.service.upstart` can
   1. use `files_switch(["consul_upstart.jinja"])`, then TOFS will select the file based on it's configuration (`["id", "os_family"]` by default)
   2. use a parameter like `upstart_init_file_path` which can be overwritten per `os_family` etc, the normal way

For things like that, I may prefer to have parameters like:

``` yaml
sysvinit:
  service_path: …
  service_name: …
  
upstart:
  config_path: …
  config_name: …

systemd:
  service_path: …
  service_name: …
```

This permit to pass a sub dictionary as `context` to `file.managed`.

Reply:
> I think the same could be think of about `init` systems, in your PR, only `service_path` is configured depending of the `init` system, `parameter/init/` could disappear completely, see below.
> 
> >
@baby-gnu @myii
I agree with that, but I think this is a lack a logic in Saltstack itself:
- The `init` system provided in FreeBSD by default is neither `systemd` nor  `sysVinit `  nor `rc`. It is a specific `init` system that exists only in FreeBSD.
- Another strange thing in Saltstack by the way is that `salt-call test.provider ` command returns `freebsd` and spots the fact that `init` system in `FreeBSD` is unique and specific for `FreeBSD` systems

Maybe the proper solution is to propose a Pull Request to Saltstack core for providing a `grains['init']` for `FreeBSD` systems

Reply:
> @baby-gnu @myii I agree with that, but I think this is a lack a logic in Saltstack itself:
> 
>     * The `init` system provided in `FreeBSD` by default is neither `systemd` nor  `sysVinit` nor `rc`. It is a specific `init` system that exists only in `FreeBSD`.
> 
>     * Another strange thing in Saltstack by the way is `salt-call test.provider service` command returns `freebsdservice` and spots the fact that `init` system in `FreeBSD` is unique and specific for `FreeBSD` systems
>       (cf. our PMs on `saltstack-formula`'s Slack channel )
> 
> Maybe the proper solution is to propose a Pull Request to Saltstack core for providing a `grains['init']` for `FreeBSD` systems

For sure, in the mean time, `consul.service` could use `salt-call test.provider service` to select which `consul.service.{{ init_system }}` to include:

``` sls
# -*- coding: utf-8 -*-
# vim: ft=sls

{#- Get the `tplroot` from `tpldir` #}
{%- set tplroot = tpldir.split('/')[0] %}
{%- from tplroot ~ "/map.jinja" import mapdata as consul with context %}

{%- set init_system = grains.get("init", False) %}
{%- if not init_system %}
      {%- set init_system = salt["test.provider"]("service") %}
{%- endif %}

include:
   - .{{ init_system }}
```

Reply:
Yep, I can do that but there is no different states to run following which `init` system we are using yet.
I just want this to PR to refactor existing behavior whithout introducing `BREAKING CHANGES` to be able to use `TOFS` and `map.jinja` v5 properly.

Maybe your proposition can be useful for further PRs, thanks for sharing.

Using `grains.init` to select right file and fallback on `os_family` to be able to handle `FreeBSD` is working right now.
The same way for `map.jinja` and do not introduce `BREAKING CHANGES`.

The real question on this review is: 
"Is `grains['init']` more general than `grains['os'] and `grains['os_family']`"

I would believe that `grains.init` is more specific than `grains.os_family`, why ?

The main raison is actually this file (https://github.com/saltstack/salt/blob/master/salt/grains/core.py) uses 
 `grains.os_family` to determine `grains.init` and not the opposite. 
Conclusion:  `grains.init` depends on `grains.os_family`  in `saltstack` core code base

To pick up on what you said:
> I'm not sure about this, in general:
> 
> 1. software provides defaults in their code base
> 2. the OS could override these defaults at compile time

and correlate that with what is intended a formula, I would write that:
1. An OS comes with an `init` system bundled by default.
2. User can overide default `init` system with another.

Another reason that makes me believe that `grains.init` is more specific than `grains.os_family and also `grains.os`
@myii @baby-gnu What tool or documention can we use to make a decision ?

Reply:
@sticky-note My original response was in respect to where you added `init` in `map_jinja.yaml`.  Considering the discussion that has taken place since, then `init` appears to be more specific than `osfinger`, even -- with `Gentoo-2` being an existing example.  Meaning:

```yaml
values:
  sources:
    - "Y:G@cpuarch"  # Default is "Y:G@osarch"
    - "Y:G@os_family"
    - "Y:G@os"
    - "Y:G@osfinger"
    - "Y:G@init"
    - "C@consul:lookup"
    - "C@consul"
    - "Y:G@id"
```

Reply:
Wonderful, I've force pushed the requested changes. @myii @baby-gnu